src/Cms/License.php

```diff
@@ -0,0 +1,512 @@
+	 */
+	public function isOnCorrectDomain(): bool
+	{
+		if ($this->domain === null) {
+			return false;
+		}
+
+		// compare domains
+		if ($this->sanitizeDomain(App::instance()->system()->indexUrl()) !== $this->sanitizeDomain($this->domain)) {
+			return false;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Compares the signature with all ingredients
+	 */
+	public function isSigned(): bool
+	{
+		if ($this->signature === null) {
+			return false;
+		}
+
+		// get the public key
+		$pubKey = F::read(App::instance()->root('kirby') . '/kirby.pub');
+
+		// verify the license signature
+		$data      = json_encode($this->signatureData());
+		$signature = hex2bin($this->signature);
+
+		if (openssl_verify($data, $signature, $pubKey, 'RSA-SHA256') !== 1) {
+			return false;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Returns a reliable label for the license type
+	 */
+	public function label(): string
+	{
+		if ($this->status() === LicenseStatus::Missing) {
+			return LicenseType::Invalid->label();
+		}
+
+		return $this->type()->label();
+	}
+
+	/**
+	 * Returns the order id if available
+	 */
+	public function order(): string|null
+	{
+		return $this->order;
+	}
+
+	/**
+	 * Support the old license file dataset
+	 * from older licenses
+	 */
+	public static function polyfill(array $license): array
+	{
+		return [
+			'activation' => $license['activation'] ?? null,
+			'code'       => $license['code']       ?? $license['license'] ?? null,
+			'date'       => $license['date']  	   ?? null,
+			'domain'     => $license['domain']     ?? null,
+			'email'      => $license['email']      ?? null,
+			'order'      => $license['order']      ?? null,
+			'signature'  => $license['signature']  ?? null,
+		];
+	}
+
+	/**
+	 * Reads the license file in the config folder
+	 * and creates a new license instance for it.
+	 */
+	public static function read(): static
+	{
+		try {
+			$license = Json::read(App::instance()->root('license'));
+		} catch (Throwable) {
+			return new static();
+		}
+
+		return new static(...static::polyfill($license));
+	}
+
+	/**
+	 * Sends a request to the hub to register the license
+	 */
+	public function register(): static
+	{
+		if ($this->type() === LicenseType::Invalid) {
+			throw new InvalidArgumentException(['key' => 'license.format']);
+		}
+
+		if ($this->hasValidEmailAddress() === false) {
+			throw new InvalidArgumentException(['key' => 'license.email']);
+		}
+
+		if ($this->domain === null) {
+			throw new InvalidArgumentException(['key' => 'license.domain']);
+		}
+
+		$response = $this->request('register', [
+			'license' => $this->code,
+			'email'   => $this->email,
+			'domain'  => $this->domain
+		]);
+
+		return $this->update($response);
+	}
+
+	/**
+	 * Returns the renewal date
+	 */
+	public function renewal(string|IntlDateFormatter|null $format = null): int|string|null
+	{
+		if ($this->activation === null) {
+			return null;
+		}
+
+		$time = strtotime('+3 years', $this->activation());
```

Comment:
Maybe it would make sense to add the renewal date to the license data. Otherwise we could never change the duration in the future because old versions would still have the hardcoded three years.

Reply:
I generally agree, but looking at how brittle it is to get the verification right on both ends, every bit of information that does not have to be in the license file is better for us. I think we can easily solve this in later releases.

Reply:
You are right. I also went through the possible cases in my head and actually there is no case where we the hardcoded period in an older Kirby version could break the licensing. We can only change the duration for newly purchased licenses and the renewal date will always be after the purchase. So A Kirby version released before any change to the duration of the period won't care about the change.

And in the worst case, we could change the license code prefix, which will prevent activation in older releases that don't yet have the required change.